### DISCUSSION TITLE
What are the schedulers and queues exactly?

### DISCUSSION THREAD

**Discussion by vonWojtas** [vernemq/vernemq#2042] [Q&A] [answered]
Hello everbody :)  

I've get a few questions to understand the metrics and system correctly. I am sorry when there is documentation but I did not find any about those topics. So please bear with me 🐻 

Our setup is currently only one node with the single application :) 

What are schedulers?
Apparently we got 4 schedulers running. One with high 100% utilization, the others are more or less inactive:
![image](https://user-images.githubusercontent.com/70715766/194000130-a093fe04-41cf-4e9f-b1fd-f7a23a6c702b.png)

So I kind of get the idea of a usual OS scheduler, telling the CPU what task to perform, but what are those in the sence of VerneMQ?

Another question which pops into my head:
What are VerneMQ's `queues`? What are `queue_processes`? And what does it mean, that I have messages in my queue?

Thank you very  much and have a nice day!

**Comment by ioolkos** [comment, marked as answer]
@vonWojtas Schedulers are related to the Erlang runtime. In most cases, you'll see 1 scheduler per CPU. If you're interested in the details, read up on Erlang scheduling.

Queues are not related to schedulers. A queue is a process that always belongs to an MQTT consumer exclusively. Every queue in VerneMQ belongs to exactly 1 MQTT consumer, and every MQTT consumer has exactly 1 queue. If you have 10k consumers on a broker, you'll have 10k queue processes. There is no concurrent destructive consuming with multiple consumers from the same queue, as with other messaging protocols.

A queue will receive every message that a consumer needs (based on the subscriptions the consumer has issued).

Note: in some sort, queues are an implementation detail. MQTT (the protocol) is queue-less! Most importantly, a topic is not a queue.

**Reply by vonWojtas** [reply]
@ioolkos thank you very much for your quick answer!

**Comment by vonWojtas** [comment]
Just for completeness sake:
> The Scheduler is responsible for choosing the Erlang process to execute. Basically the scheduler keeps two queues, a ready queue of processes ready to run, and a waiting queue of processes waiting to receive a message. When a process in the waiting queue receives a message or gets a time out it is moved to the ready queue.

Source: https://blog.stenmans.org/theBeamBook/#_scheduling

**Reply by ioolkos** [reply]
Correct. This is all about the implementation of preemptive concurrency. Those scheduler queues are not VerneMQ's queues, to be clear.

**Reply by vonWojtas** [reply]
Let me rephrase to help understand me better (and hopefully someone else too)
VerneMQ's queues are simple queues for every subscriber, so that those messages can be send. 
It's still important that the MQTT protocol is not a queuing protocol (If you need that, you should probably switch to AMQP).

**Reply by ioolkos** [reply]
MQTT is a PUB/SUB protocol with message persistance guarantees but no requirement to implement those as queues.
In VerneMQ, delivery and message persistence is implemented with a per-consumer queue. In AMQP, you can attach multiple consumers to a queue, in VerneMQ you cannot.

Note that a queue's job is not at all about message sending. It's about message consuming. If an MQTT client sends a message, it does not send it to any queue. It sends it to the broker directly, giving a routing key (a topic name). The broker then simply matches that routing key to existing subscriptions and finds all the queues that need that message.
